pywbem_mock/_baserepository.py

```diff
@@ -0,0 +1,395 @@
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+#
+# Author: Karl  Schopmeyer <inovadevelopment.com>
+#
+
+
+"""
+Base classes for an object store for CIM classes, CIM instances, and CIM
+qualifier declarations and the generic API for a CIM repository of these object
+types organized by repository.
+"""
+
+from abc import abstractmethod, abstractproperty
+from six import add_metaclass, PY2
+from custom_inherit import DocInheritMeta
+
+
+def compatibleabstractproperty(func):
+    """
+    Python 2 and python 3 differ in decorator for abstract property.
+    in python 3 (gt 3.3) it is:
+        @property
+        @abstractproperty
+    in python 2
+        @abstractproperty
+    """
+
+    if PY2:  # pylint: disable=no-else-return
+        return abstractproperty(func)
+    else:
+        return property(abstractmethod(func))
+
+
+@add_metaclass(DocInheritMeta(style="google", abstract_base_class=True))
+class BaseObjectStore(object):
+    """
+    An abstract class that defines the APIs for the methods of an object store
+    for CIM objects including CIMClass, CIMInstance, and CIMQualifierDeclaration
+    objectsthat constitute a WBEM server repository.  This
+    class provides the abstract methods for creating, accessing, and deleting,
+    CIM objects of a single CIM object type in the repository.
+    """
+
+    def __init__(self, cim_object_type):
+        """
+        Parameters:
+
+            cim_object_type(:term:`CIM object`):
+              The Pywbem cim object type as defined in cim_types.py for the
+              objects in the data store. Used to verify values on create.
+        """
+        self._cim_object_type = cim_object_type
+
+    @abstractmethod
+    def exists(self, name):
+        """
+        Test if CIM object defined by name exists in the object store.
+
+        Parameters:
+
+          name(:term:`string` or :class:`~pywbem.CIMInstanceName`):
+            Name by which the object is identified in the object store
+
+        Returns:
+
+            True if the object exists; False if it does not exist
+        """
+        pass
+
+    @abstractmethod
+    def get(self, name, copy=True):
+        """
+        Return the CIM object defined by name if it exists in the object store.
+
+        Parameters:
+
+          name(:term:`string` or :class:`~pywbem.CIMInstanceName`):
+            Name by which the object is identified in the object store
+
+          copy: If True, returns a copy of the object to insure that modifying
+          the returned object does not change the data store. The default
+          behavior is True .  If copy is False, the object in the object store
+          is returned but if it is modified by the user, the object in the
+          store may be modified also.
```

Comment:
This formatting leads to screwed up docs output. To fix it, the type needs to be added in parenthesis after the parameter name, and the second to last lines need to be indented relative to the first line.

Reply:
DONE

Reply:
DONE

Reply:
Not done - This change did not make it into the last commit.

Reply:
DONE in new commit.